crates/core/tedge_write/src/bin.rs

```diff
@@ -0,0 +1,160 @@
+    // the files we write to, e.g:
+    //
+    // `tedge    ALL = (ALL) NOPASSWD: /usr/bin/tedge-write /etc/*`
+    //
+    // If the destination path contains `..` then we can "escape" outside the directory we're
+    // allowed to write to. For that reason, we require paths to be canonical.
+    //
+    // Ideally this would be solved by a more expressive filesystem permissions system, e.g. ACL,
+    // but unfortunately they're not standard on Linux, so we're stuck with trying to do next best
+    // thing with sudo.
+    if !args.destination_path.is_absolute() {
+        bail!("Destination path has to be absolute");
+    }
+
+    let target_filepath: Utf8PathBuf = path_clean::clean(args.destination_path.as_str()).into();
+
+    if target_filepath != *args.destination_path {
+        bail!(
+            "Destination path {} is not canonical",
+            args.destination_path
+        );
+    }
+
+    let file_existed_before_write = target_filepath.is_file();
+
+    write_stdin_to_file_atomic(&target_filepath)?;
+
+    if file_existed_before_write {
+        return Ok(());
+    }
+
+    if args.user.is_some() || args.group.is_some() {
+        chown_by_user_and_group_name(
+            &target_filepath,
+            args.user.as_deref(),
+            args.group.as_deref(),
+        )
+        .context("Changing ownership of destination file failed")?;
+    }
+
+    if let Some(mode) = args.mode {
+        let mode = u32::from_str_radix(&mode, 8).context("Parsing mode failed")?;
+        let permissions = fs::Permissions::from_mode(mode);
+        fs::set_permissions(args.destination_path.as_std_path(), permissions)
+            .context("Could not set new permissions")?;
+    }
+
+    Ok(())
+}
+
+/// Writes contents of stdin into a file atomically.
+///
+/// To write a file atomically, stdin is written into a temporary file, which is then renamed into the target file.
+/// Because rename is only atomic if both source and destination are on the same filesystem, the temporary file is
+/// located in the same directory as the target file.
+///
+/// Using [`io::copy`], data should be copied using Linux-specific syscalls for copying between file descriptors,
+/// without unnecessary copying to and from a userspace buffer.
+fn write_stdin_to_file_atomic(target_filepath: &Utf8Path) -> anyhow::Result<()> {
+    let temp_filepath = {
+        let Some(temp_filename) = target_filepath.file_name().map(|f| format!("{f}.tmp")) else {
+            bail!("Destination path {target_filepath} does not name a valid filename");
+        };
+        target_filepath.with_file_name(temp_filename)
+    };
+
+    let mut temp_file = fs::OpenOptions::new()
+        .write(true)
+        .create_new(true)
+        .open(temp_filepath.as_std_path())
+        .with_context(|| {
+            format!(
+                "Could not open temporary file `{}` for writing",
+                temp_filepath
+            )
+        })?;
```

Comment:
Minor:
```suggestion
        .with_context(|| format!("Could not open temporary file `{temp_filepath}` for writing"))?;
```

No idea off the top of my head whether it will format to be one line or three, but it would be consistent with all the other format strings here.